api/action/types.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2014 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package action
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	"github.com/juju/juju/apiserver/params"
+)
+
+// Operations holds a list of operations and whether the
+// list has been truncated (used for batch queries).
+type Operations struct {
+	Operations []Operation
+	Truncated  bool
+}
+
+// Operation holds a list of actions run as part of the operation
+type Operation struct {
+	ID        string
+	Summary   string
+	Enqueued  time.Time
+	Started   time.Time
+	Completed time.Time
+	Status    string
+	Actions   []ActionResult
+	Error     error
+}
+
+// ActionMessage represents a logged message on an action.
+type ActionMessage struct {
+	Timestamp time.Time
+	Message   string
+}
+
+// Action is a named task to execute on a unit or machine.
+type Action struct {
+	ID         string
+	Receiver   string
+	Name       string
+	Parameters map[string]interface{}
+}
+
+// ActionResult is the result of running an action.
+type ActionResult struct {
+	*Action
```

Comment:
this is inherently dangerous, if Action is optional, which it looks like it is, embedding Action could cause a programming error where code doesn't nil check the embedded Action, instead might just nil check a ptr to ActionResult.

Two ways to fix this, Action is embedded but not a ptr or Action is a property rather than embedded.

Reply:
changed to *Action as an attribute